src/include/type/expression/case_expression.h

```diff
@@ -0,0 +1,128 @@
+#pragma once
+
+#include <memory>
+#include <utility>
+#include <vector>
+#include "type/expression/abstract_expression.h"
+
+namespace terrier {
+namespace type {
+namespace expression {
+
+/**
+ * Represents a logical case expression.
+ */
+class CaseExpression : public AbstractExpression {
+ public:
+  /**
+   * A unique pointer to an abstract expression.
+   */
+  using AbsExprPtr = std::unique_ptr<AbstractExpression>;
+  /**
+   * WHEN ... THEN ... clauses.
+   */
+  using WhenClause = struct {
+    /**
+     * The condition to be checked for this case expression.
+     */
+    AbsExprPtr condition;
+    /**
+     * The value that this expression should have if the corresponding condition is true.
+     */
+    AbsExprPtr then;
+  };
+
+  /**
+   * Instantiate a new case expression.
+   * @param return_value_type return value of the case expression
+   * @param when_clauses list of when clauses
+   * @param default_expr default expression for this case
+   */
+  explicit CaseExpression(const TypeId return_value_type, std::vector<WhenClause> &&when_clauses,
```

Comment:
r-value references here seem odd to me. I'd add a TODO to revisit this once we actually use it to see if that signature makes sense. My gut says no.